c_src/enif_jq.c

```diff
@@ -65,59 +83,82 @@ static ERL_NIF_TERM make_ok_return(ErlNifEnv* env, ERL_NIF_TERM result) {
 
 static ERL_NIF_TERM parse_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
 {
-    char jv_filter[MAX_JQ_FILTER_LEN] = {0};
+    ERL_NIF_TERM error_msg_bin; 
     // ----------------------------- init --------------------------------------
     jq_state *jq = NULL;
     int ret = JQ_ERROR_UNKNOWN;
     ERL_NIF_TERM ret_term;
-    memset(err_msg, 0, MAX_ERR_MSG_LEN * sizeof(char));
     int dumpopts = 512; // JV_PRINT_SPACE1
     jq = jq_init();
     if (jq == NULL) {
         ret = JQ_ERROR_SYSTEM;
-        strcpy(err_msg, "jq_init: malloc error");
+        const char* error_message = "jq_init: Could not initialize jq";
+        error_msg_bin =
+            make_error_msg_bin(env, error_message, strlen(error_message));
         goto out;
     }
-    jq_set_error_cb(jq, err_callback, err_msg);
+    NifEnvAndErrBinPtr env_and_msg_bin = {
+        .env = env,
+        .error_msg_bin_ptr = &error_msg_bin
+    };
+    jq_set_error_cb(jq, err_callback, &env_and_msg_bin);
 
     // --------------------------- read args -----------------------------------
-    ErlNifBinary erl_jq_filter, erl_json_text;
+    ErlNifBinary erl_jq_filter;
+    ErlNifBinary erl_json_text;
     if (!enif_inspect_binary(env, argv[0], &erl_jq_filter) ||
         !enif_inspect_binary(env, argv[1], &erl_json_text)) {
         ret = JQ_ERROR_BADARG;
-        strcpy(err_msg, "invalid input or filter");
+        const char* error_message = "Expected arguments of type binary but got something else";
+        error_msg_bin =
+            make_error_msg_bin(env, error_message, strlen(error_message));
         goto out;
-	}
+    }
 
     // ------------------------- parse input json -----------------------------
-    jv jv_json_text = jv_parse_sized((const char*)erl_json_text.data, erl_json_text.size);
+    // It is reasonable to assume that jv_parse_sized would not require the 
+    // input string to be NULL terminated. Unfortunately this is not the case.
+    // Error reporting depends on that the input string is NULL terminated
+    // so we have to copy the input to a larger memory block to make sure it
```

Comment:
In case we are going to support iolist inputs, it maybe easier just append '\0' at the end of iolist.

Reply:
Yes, but the input still need to be converted to a continues block of memory to be passed to the jq library so we would need to create a new block even if we support iolists.